lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java

```diff
@@ -0,0 +1,171 @@
+        //it to instantiate the actual runtime.
+        Injector rootInjector = Tang.Factory.getTang().newInjector();
+        initializeInjector(rootInjector);
+        final JavaConfigurationBuilder cb = Tang.Factory.getTang().newConfigurationBuilder();
+        cb.bindNamedParameter(RuntimeName.class, rd.getRuntimeName().toString());
+        cb.bindImplementation(Runtime.class, RuntimeImpl.class);
+
+        AvroConfigurationSerializer serializer = new AvroConfigurationSerializer();
+        Configuration config = serializer.fromString(rd.getSerializedConfiguration().toString());
+        final Injector runtimeInjector = rootInjector.forkInjector(config, cb.build());
+        this.runtimes.put(rd.getRuntimeName().toString(), runtimeInjector.getInstance(Runtime.class));
+      } catch (InjectionException e) {
+        throw new RuntimeException("Unable to initialize runtimes.", e);
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to initialize runtimes.", e);
+      }
+    }
+  }
+
+  /**
+   * Initializes injector by copying needed handlers.
+   * @param runtimeInjector Teh injector to initialize
+   * @throws InjectionException
+   */
+  private void initializeInjector(final Injector runtimeInjector) throws InjectionException {
+    final EventHandler<ResourceStatusEvent> statusEventHandler =
+            this.originalInjector.getNamedInstance(RuntimeParameters.ResourceStatusHandler.class);
+    runtimeInjector.bindVolatileParameter(RuntimeParameters.ResourceStatusHandler.class, statusEventHandler);
+    final EventHandler<NodeDescriptorEvent> nodeDescriptorEventHandler =
+            this.originalInjector.getNamedInstance(RuntimeParameters.NodeDescriptorHandler.class);
+    runtimeInjector.bindVolatileParameter(RuntimeParameters.NodeDescriptorHandler.class, nodeDescriptorEventHandler);
+    final EventHandler<ResourceAllocationEvent> resourceAllocationEventHandler =
+            this.originalInjector.getNamedInstance(RuntimeParameters.ResourceAllocationHandler.class);
+    runtimeInjector.bindVolatileParameter(
+            RuntimeParameters.ResourceAllocationHandler.class,
+            resourceAllocationEventHandler);
+    final EventHandler<RuntimeStatusEvent> runtimeStatusEventHandler =
+            this.originalInjector.getNamedInstance(RuntimeParameters.RuntimeStatusHandler.class);
+    runtimeInjector.bindVolatileParameter(
+            RuntimeParameters.RuntimeStatusHandler.class,
+            runtimeStatusEventHandler);
+  }
+
+  /**
+   * Retrieves requested runtime, if requested name is empty a default runtime will be used.
+   * @param requestedRuntimeName teh requested runtime name
+   * @return
+   */
+  private Runtime getRuntime(final String requestedRuntimeName) {
+    String runtimeName = requestedRuntimeName;
+    if (StringUtils.isEmpty(runtimeName) || StringUtils.isBlank(runtimeName)) {
+      runtimeName = this.defaultRuntimeName;
+    }
+
+    return this.runtimes.get(runtimeName);
```

Comment:
Not quite. If the string isn't blank (which is the only thing you need to check, as blank implies empty), it can still refer to a runtime that isn't in `this.runtimes`. In other words:

```java
    return this.runtimes.get(runtimeName);
```

could return `null` with no explanation why.